File: javascript/src/transform.js
import { KvlConfig } from './config.js';

/**
 * Merge two categorical models recursively.
 *
 * When both are dicts → recursive merge.
 * Conflicting scalars → convert to categorical and merge.
 *
 * @param {*} model1
 * @param {*} model2
 * @returns {*}
 */
export function merge(model1, model2) {
  if (isDict(model1) && isDict(model2)) {
    const result = { ...model1 };
    for (const [k, v] of Object.entries(model2)) {
      result[k] = k in result ? merge(result[k], v) : v;
    }
    return result;
  }

  if (isDict(model1)) return model1;
  if (isDict(model2)) return model2;

  // Both non-dict → convert to categorical and merge
  const cat1 = model1 != null && model1 !== '' ? { [String(model1)]: {} } : {};
  const cat2 = model2 != null && model2 !== '' ? { [String(model2)]: {} } : {};
  return merge(cat1, cat2);
}

/**
 * Compact KVL structures:
 *  1. Singleton empty key lifting: {'': content} → content
 *  2. Single empty child → string: {'value': {}} → 'value'
 *  3. All-empty-dict flattening: {'a': {}, 'b': {}} → ['a', 'b']
 *
 * @param {*} data
 * @param {KvlConfig} [config]
 * @returns {*}
 */
export function compact(data, config) {
  config = config ?? new KvlConfig();

  if (Array.isArray(data)) {
    return data.map(item => compact(item, config));
  }
  if (!isDict(data)) return data;

  // Recursively compact nested structures first
  const compacted = {};
  for (const [key, value] of Object.entries(data)) {
    compacted[key] = compact(value, config);
  }

  return applyCompactingRules(compacted, config);
}

/**
 * Expand lists back to categorical dict format for serialization.
 * Inverse of compact().
 *
 * @param {*} data
 * @returns {*}
 */
export function expand(data) {
  if (Array.isArray(data)) {
    const result = {};
    for (const item of data) {
      if (typeof item === 'string') {
        result[item] = {};
      } else {
        const expanded = expand(item);
        if ('' in result) {
          result[''] = merge(result[''], expanded);
        } else {
          result[''] = expanded;
        }
      }
    }
    return result;
  }

  if (isDict(data)) {
    const result = {};
    for (const [key, value] of Object.entries(data)) {
      result[key] = expand(value);
    }
    return result;
  }

  // Primitive → categorical
  if (data == null) return {};
  return { [String(data)]: {} };
}

/**
 * @param {object} data
 * @param {KvlConfig} config
 * @returns {*}
 */
function applyCompactingRules(data, config) {
  if (!isDict(data) || Object.keys(data).length === 0) return data;

  const keys = Object.keys(data);

  // Rule 1: Singleton empty key lifting
  if (keys.length === 1 && keys[0] === '') {
    const emptyValue = data[''];
    if (
      isDict(emptyValue) &&
      Object.keys(emptyValue).length > 0 &&
      Object.values(emptyValue).every(v => isDict(v))
    ) {
      return Object.entries(emptyValue).map(([k, v]) => ({ [k]: v }));
    }
    return emptyValue;
  }

  // Rule 2: Single empty child → string
  if (keys.length === 1) {
    const childKey = keys[0];
    const childValue = data[childKey];
    if (isDict(childValue) && Object.keys(childValue).length === 0 && childKey !== '') {
      return childKey;
    }
  }

  // Rule 3: All-empty-dict flattening to list
  if (Object.values(data).every(v => isDict(v) && Object.keys(v).length === 0)) {
    // Rule 4: Don't flatten if any key looks like a section header
    for (const key of keys) {
      if (typeof key === 'string' && isSectionHeader(key, config.separator)) {
        return data;
      }
    }
    return keys;
  }

  return data;
}

/**
 * Check if a value looks like a section header.
 * @param {string} value
 * @param {string} separator
 * @returns {boolean}
 */
function isSectionHeader(value, separator) {
  if (typeof value !== 'string' || !value.trim()) return false;
  return value.trim().startsWith(separator);
}

/**
 * @param {*} value
 * @returns {boolean}
 */
function isDict(value) {
  return value !== null && typeof value === 'object' && !Array.isArray(value);
}

File: rust/src/transform.rs
//! Structural transformations: categorical merge, compacting, and expansion.

use crate::config::KvlConfig;
use crate::value::{Map, Value};

/// Merge two categorical models recursively.
///
/// Maps merge key-by-key; a map wins over a non-map; two conflicting scalar
/// values are lifted into categorical form and merged, so
/// `port=8080 + port=8081` becomes `{port: {8080: {}, 8081: {}}}`.
pub fn merge(a: &Value, b: &Value) -> Value {
    match (a, b) {
        (Value::Map(m1), Value::Map(m2)) => {
            let mut result = m1.clone();
            for (k, v) in m2 {
                match result.get(k) {
                    Some(existing) => {
                        let merged = merge(existing, v);
                        result.insert(k.clone(), merged);
                    }
                    None => {
                        result.insert(k.clone(), v.clone());
                    }
                }
            }
            Value::Map(result)
        }
        (Value::Map(_), _) => a.clone(),
        (_, Value::Map(_)) => b.clone(),
        _ => {
            let cat1 = to_categorical_scalar(a);
            let cat2 = to_categorical_scalar(b);
            merge(&cat1, &cat2)
        }
    }
}

fn to_categorical_scalar(v: &Value) -> Value {
    match v {
        Value::String(s) if s.is_empty() => Value::empty(),
        Value::String(s) => {
            let mut m = Map::new();
            m.insert(s.clone(), Value::empty());
            Value::Map(m)
        }
        Value::List(items) => {
            let mut result = Value::empty();
            for item in items {
                result = merge(&result, &to_categorical_scalar(item));
            }
            result
        }
        Value::Map(_) => v.clone(),
    }
}

/// Compact a categorical structure into the user-facing form:
///
/// 1. Singleton empty-key lifting: `{"": content}` becomes `content`
/// 2. Single empty child conversion: `{"value": {}}` becomes `"value"`
/// 3. Empty-value map flattening: `{"a": {}, "b": {}}` becomes `["a", "b"]`
/// 4. Section-header keys (starting with the separator) are never flattened
pub fn compact(data: &Value, config: &KvlConfig) -> Value {
    match data {
        Value::List(items) => Value::List(items.iter().map(|i| compact(i, config)).collect()),
        Value::String(_) => data.clone(),
        Value::Map(m) => {
            let compacted: Map = m
                .iter()
                .map(|(k, v)| (k.clone(), compact(v, config)))
                .collect();
            apply_compacting_rules(compacted, config)
        }
    }
}

fn apply_compacting_rules(data: Map, config: &KvlConfig) -> Value {
    if data.is_empty() {
        return Value::Map(data);
    }

    // Rule 1: singleton empty-key lifting
    if data.len() == 1 && data.contains_key("") {
        let empty_value = &data[""];
        if let Value::Map(ev) = empty_value {
            if !ev.is_empty() && ev.values().all(|v| matches!(v, Value::Map(_))) {
                return Value::List(
                    ev.iter()
                        .map(|(k, v)| {
                            let mut m = Map::new();
                            m.insert(k.clone(), v.clone());
                            Value::Map(m)
                        })
                        .collect(),
                );
            }
        }
        return empty_value.clone();
    }

    // Rule 2: single empty child converts to a string value
    if data.len() == 1 {
        let (key, value) = data.iter().next().unwrap();
        if value.is_empty_map() && !key.is_empty() {
            return Value::String(key.clone());
        }
    }

    // Rule 3: all-empty-value maps flatten to a list of keys
    if data.values().all(|v| v.is_empty_map()) {
        // Rule 4: keep maps whose keys look like section headers
        if data
            .keys()
            .any(|k| is_section_header(k, &config.separator))
        {
            return Value::Map(data);
        }
        return Value::List(data.keys().map(|k| Value::String(k.clone())).collect());
    }

    Value::Map(data)
}

fn is_section_header(value: &str, separator: &str) -> bool {
    let stripped = value.trim();
    !stripped.is_empty() && stripped.starts_with(separator)
}

/// Expand compacted data back to the categorical nested-map structure.
/// This is the inverse of `compact()`, used before categorical serialization.
pub fn expand(data: &Value) -> Value {
    match data {
        Value::List(items) => {
            let mut result = Map::new();
            for item in items {
                if let Value::String(s) = item {
                    result.insert(s.clone(), Value::empty());
                } else {
                    let expanded = expand(item);
                    let existing = result
                        .entry(String::new())
                        .or_insert_with(Value::empty)
                        .clone();
                    result.insert(String::new(), merge(&existing, &expanded));
                }
            }
            Value::Map(result)
        }
        Value::Map(m) => Value::Map(m.iter().map(|(k, v)| (k.clone(), expand(v))).collect()),
        Value::String(s) => {
            if s.contains('\n') {
                data.clone()
            } else {
                let mut m = Map::new();
                m.insert(s.clone(), Value::empty());
                Value::Map(m)
            }
        }
    }
}
